File: python/puzzle.py
import multiprocessing
import threading
import copy
import heapq
import os
import sys
import time
from enum import Enum
from math import sqrt, floor
from typing import Optional

class Action(Enum):
    none = 1
    up = 2
    down = 3
    left = 4
    right = 5

    def __str__(self):
        match self:
            case self.up:
                return "Up"
            case self.down:
                return "Down"
            case self.left:
                return "Left"
            case self.right:
                return "Right"
            case _:
                return "None"


Position = tuple[int, int]


def index_of_pos(pos: Position, n: int) -> int:
    return pos[0] * n + pos[1]


def pos_of_index(i: int, n: int) -> Position:
    return i // n, i % n


def in_bounds(pos: Position, n: int) -> bool:
    return 0 <= pos[0] < n and 0 <= pos[1] < n


def add_positions(pos1: Position, pos2: Position) -> Position:
    return pos1[0] + pos2[0], pos1[1] + pos2[1]


Direction = tuple[Position, Action]
directions: list[Direction] = [((1, 0), Action.down), ((-1, 0), Action.up), ((0, 1), Action.right), ((0, -1), Action.left)]


class Puzzle:
    def __init__(self, tiles, n):
        self.g = 0
        self.f = 0
        self.prev = None
        self.tiles = tiles
        self.action = Action.none
        self.n = n

    def equals(self, other: list[int]) -> bool:
        for i in range(0, len(self.tiles)):
            if self.tiles[i] != other[i]:
                return False
        return True

    def heuristic(self) -> int:
        h = 0
        for i in range(0, len(self.tiles)):
            tile = self.tiles[i]
            if tile != 0:
                row1, col1 = pos_of_index(i, self.n)
                row2, col2 = pos_of_index(self.tiles[i], self.n)
                h += abs(row2 - row1) + abs(col2 - col1)
        return h

    def find_zero(self) -> Position:
        for i in range(0, len(self.tiles)):
            if self.tiles[i] == 0:
                return pos_of_index(i, self.n)
        raise Exception("Puzzles should contain a 0 tile")

    def to_string(self):
        s = ''
        for tile in self.tiles:
            s += str(tile)
        return s

    def print(self):
        for i in range(0, len(self.tiles)):
            if self.tiles[i] == 0:
                print("  ", end="")
            else:
                print(self.tiles[i], end=" ")
            if (i + 1) % self.n == 0:
                print("\n", end="")

    def __lt__(self, other):
        return self.f < other.f


def create_goal(size: int):
    tiles = []
    for i in range(0, size):
        tiles.append(i)
    return tiles


def reconstruct_path(curr: Optional[Puzzle]):
    path = []
    while curr is not None:
        path.append(curr)
        curr = curr.prev
    path.reverse()
    return path


def find_path(initial: Puzzle):
    start = time.perf_counter()

    visited = {}
    frontier = [initial]

    goal = create_goal(len(initial.tiles))

    path = []

    nodes = 0
    while len(frontier) > 0:
        curr_puzzle = heapq.heappop(frontier)
        nodes += 1

        if curr_puzzle.equals(goal):
            path = reconstruct_path(curr_puzzle)
            break

        visited[curr_puzzle.to_string()] = True

        zero_pos = curr_puzzle.find_zero()
        zero_index = index_of_pos(zero_pos, curr_puzzle.n)

        for direction in directions:
            next_pos = add_positions(zero_pos, direction[0])
            next_index = index_of_pos(next_pos, curr_puzzle.n)

            if not in_bounds(next_pos, curr_puzzle.n):
                continue

            next_puzzle = Puzzle(copy.deepcopy(curr_puzzle.tiles), curr_puzzle.n)

            temp = next_puzzle.tiles[zero_index]
            next_puzzle.tiles[zero_index] = next_puzzle.tiles[next_index]
            next_puzzle.tiles[next_index] = temp

            next_puzzle.prev = curr_puzzle
            next_puzzle.action = direction[1]
            next_puzzle.g = curr_puzzle.g + 1
            next_puzzle.f = next_puzzle.g + next_puzzle.heuristic()

            if not visited.get(next_puzzle.to_string(), False):
                heapq.heappush(frontier, next_puzzle)

    end = time.perf_counter()
    t = (end - start) * 1000.0

    return path, nodes, t


def read_puzzles(s: str) -> list[Puzzle]:
    puzzles = []
    curr_tiles = []

    lines = s.split("\n")
    for line in lines:
        tokens = line.split(" ")

        if len(tokens) > 1:
            for token in tokens:
                if token != "":
                    curr_tiles.append(int(token))
        else:
            if len(curr_tiles) > 0:
                n = floor(sqrt(len(curr_tiles)))
                puzzles.append(Puzzle(curr_tiles, n))
                curr_tiles = []
    return puzzles


def run_puzzles(puzzle):
    solutions = []
    for puzzle in puzzles:
        solutions.append(find_path(puzzle))
    return solutions


def run_puzzle_task(puzzle, i, solutions):
    solutions[i] = find_path(puzzle)


def run_puzzles_parallel(puzzles):
    manager = multiprocessing.Manager()
    solutions = manager.list([None] * len(puzzles))

    processes = []
    for i in range(0, len(puzzles)):
        p = multiprocessing.Process(target=run_puzzle_task, args=(puzzles[i], i, solutions))
        processes.append(p)
        p.start()
    for p in processes:
        p.join()

    return solutions


if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Need at least 1 program argument")
        exit(1)

    flag = "seq"
    if len(sys.argv) > 2:
        flag = sys.argv[2]

    with open(sys.argv[1], "r") as file:
        file_contents = file.read()
        puzzles = read_puzzles(file_contents)

        start = time.perf_counter()

        if flag == "seq":
            solutions = run_puzzles(puzzles)
        elif flag == "par":
            solutions = run_puzzles_parallel(puzzles)
        else:
            print("Flag must be par or seq got ", flag)
            exit(1);

        end = time.perf_counter()
        eteTime = (end - start) * 1000.0

        for sol in solutions:
            for p in sol[0]:
                print(p.action)

            print(f"Solved in {len(sol[0]) - 1} steps\n")

        totalTime = 0
        totalNodes = 0
        for i in range(0, len(solutions)):
            print(f"Puzzle {i + 1}: {solutions[i][2]} ms, {solutions[i][1]} nodes")
            totalTime += solutions[i][2]
            totalNodes += solutions[i][1]

        print(f"\nTotal: {totalTime} ms, {totalNodes} nodes")
        print(f"End-to-end: {eteTime} ms")
            

File: nodejs/puzzle.js
const fs = require('fs');
const Heap = require('heap');

const ACTIONS = ["Start", "Up", "Down", "Left", "Right"]

const DIRECTIONS = [
    { row: 1, col: 0 },
    { row: -1, col: 0 },
    { row: 0, col: 1 },
    { row: 0, col: -1 }
];

function newPuzzle(tiles) {
    return { 
        g: 0,
        f: 0,
        prev: null,
        tiles: tiles,
        action: 0
    };
}

function tilesEquals(tiles, other) {
    for (let i = 0; i < tiles.length; i++) {
        if (tiles[i] !== other[i]) {
            return false;
        }
    }
    return true;
}

function heuristic(tiles, dim) {
    let h = 0;
    for (let i = 0; i < tiles.length; i++) {
        const t = tiles[i];
        if (t != 0) {
            const row1 = Math.floor(i / dim);
            const col1 = i % dim;
            const row2 = Math.floor(t / dim);
            const col2 = t % dim;
            h += Math.abs(row2 - row1) + Math.abs(col2 - col1);
        }
    }
    return h;
}

function findZero(tiles) {
    for (let i = 0; i < tiles.length; i++) {
        if (tiles[i] === 0) {
            return i;
        }
    }
    throw new Error("Puzzles should contain a 0 tile");
}

function tilesToString(tiles) {
    return tiles.join('');
}

function printString(dim) {
    let s = "";
    for (let i = 0; i < tiles.length; i++) {
        if (tiles[i] === 0) {
            s += "  ";
        } else {
            s += tiles[i] + " ";
        }
        if ((i + 1) % dim === 0) {
            s += "\n";
        }
    }
    return s
}

function inBounds(row, col, dim) {
    return row >= 0 && row < dim && col >= 0 && col < dim;
}

function createGoal(size) {
    const tiles = [];
    for (let i = 0; i < size; i++) {
        tiles.push(i);
    }
    return tiles;
}

function createGoal(size) {
    const tiles = [];
    for (let i = 0; i < size; i++) {
        tiles.push(i);
    }
    return tiles;
}


function reconstructPath(curr) {
    const path = [];
    while (curr !== null) {
        path.push(curr);
        curr = curr.prev;
    }
    path.reverse();
    return path;
}

function findPath(initial) {
    const dim = Math.floor(Math.sqrt(initial.tiles.length));

    const visited = new Map();
    const frontier = new Heap((p1, p2) => p1.f - p2.f);
    frontier.push(initial);

    const goal = createGoal(initial.tiles.length);

    let nodes = 0;
    while (!frontier.empty()) {
        const currPuzzle = frontier.pop();
        nodes++;

        if (tilesEquals(currPuzzle.tiles, goal)) {
            return { path: reconstructPath(currPuzzle), nodes };
        }

        const tilesStr = tilesToString(currPuzzle.tiles);
        visited.set(tilesStr, true);

        const zeroIndex = findZero(currPuzzle.tiles);
        const zeroRow = Math.floor(zeroIndex / dim);
        const zeroCol = zeroIndex % dim;

        for (let i = 0; i < DIRECTIONS.length; i++) {
            const {row, col} = DIRECTIONS[i];

            const nextRow = zeroRow + row;
            const nextCol = zeroCol + col;

            if (!inBounds(nextRow, nextCol, dim)) {
                continue;
            }

            const nextIndex = nextRow * dim + nextCol;
            const nextPuzzle = newPuzzle([...currPuzzle.tiles]);

            const temp = nextPuzzle.tiles[zeroIndex];
            nextPuzzle.tiles[zeroIndex] = nextPuzzle.tiles[nextIndex];
            nextPuzzle.tiles[nextIndex] = temp;

            nextPuzzle.prev = currPuzzle;
            nextPuzzle.action = i + 1;
            nextPuzzle.g = currPuzzle.g + 1;
            nextPuzzle.f = nextPuzzle.g + heuristic(nextPuzzle.tiles, dim);

            const tilesStr = tilesToString(nextPuzzle.tiles);
            if (!visited.get(tilesStr)) {
                frontier.push(nextPuzzle);
            }
        }
    }

    return { path: [], nodes };
}

function readPuzzles(s) {
    const puzzles = [];
    let currTiles = [];

    const lines = s.split("\n");
    for (const line of lines) {
        const tokens = line.split(" ").map(t => t.trim()).filter((t) => t != "");

        if (tokens.length > 1) {
            for (const token of tokens) {
                currTiles.push(parseInt(token));
            }
        } else {
            if (currTiles.length > 0) {
                puzzles.push(newPuzzle(currTiles));
                currTiles = [];
            }
        }
    }
    return puzzles;
}

function runPuzzles(puzzles) {
    const solutions = [];
    for (const puzzle of puzzles) {
        const start = process.hrtime.bigint();

        const { path, nodes } = findPath(puzzle);

        const end = process.hrtime.bigint();
        const time = Number(end - start) / 1e6;

        solutions.push({path, time, nodes});
    }
    return solutions;
}

function main() {
    if (process.argv.length < 3) {
        console.log("Need at least 1 program argument");
        process.exit(1);
    }

    const path = process.argv[2];

    let fileContents = "";
    try {
        fileContents = fs.readFileSync(path, "utf8");
    } catch (err) {
        console.log("Failed to read input file " + path)
        process.exit(1);
    }

    const puzzles = readPuzzles(fileContents);
    const solutions = runPuzzles(puzzles);

    for (const sol of solutions) {
        for (const p of sol.path) {
            console.log(ACTIONS[p.action]);
        }
        console.log(`Solved in ${sol.path.length - 1} steps\n`);
    }

    let totalTime = 0;
    let totalNodes = 0;
    for (let i = 0; i < solutions.length; i++) {
        console.log(`Puzzle ${i + 1}: ${solutions[i].time} ms, ${solutions[i].nodes} nodes`);
        totalTime += solutions[i].time;
        totalNodes += solutions[i].nodes;
    }

    console.log(`\nTotal: ${totalTime} ms, ${totalNodes} nodes`);
    console.log(`End-to-end: ${totalTime} ms`);
}

main();
